packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts

```diff
@@ -0,0 +1,127 @@
+import { UniqueIdentifier } from "@dnd-kit/core";
+import { arrayMove } from "@dnd-kit/sortable";
+import { groupBy } from "lodash";
+import { AccountPage } from "../../../../components/hooks/useAccountPages";
+
+export interface PageList {
+  treeList: TreeElement[];
+  flatList: TreeElement[];
+}
+
+export interface TreeElement extends AccountPage {
+  parentId: string;
+  depth: number;
+  index: number;
+  expanded: boolean;
+  expandable: boolean;
+  collapsed: boolean;
+  children?: TreeElement[];
+}
+
+const getRecursivePageList = (
+  groupedPages: { [id: string]: AccountPage[] },
+  id: string,
+  expandedIds: string[],
+  depth = 0,
+  collapsed = false,
+): PageList => {
+  const emptyTreeList = [] as TreeElement[];
+  const emptyList: PageList = {
+    treeList: emptyTreeList,
+    flatList: emptyTreeList,
+  };
+
+  return (
+    groupedPages[id]?.reduce((prev, page, index) => {
+      const expanded = expandedIds.includes(page.entityId);
+      const children = getRecursivePageList(
+        groupedPages,
+        page.entityId,
+        expandedIds,
+        depth + 1,
+        collapsed || !expanded,
+      );
+      const expandable = !!children.treeList.length;
+
+      const item = {
+        ...page,
+        depth,
+        index,
+        parentId: page.parentPageEntityId,
+        expanded,
+        expandable,
+        collapsed,
+      } as TreeElement;
+
+      return {
+        treeList: [
+          ...prev.treeList,
+          {
+            ...item,
+            children: children.treeList,
+          } as TreeElement,
+        ],
+        flatList: [...prev.flatList, item, ...children.flatList],
+      };
+    }, emptyList) || emptyList
+  );
+};
+
+export const getPageList = (pages: AccountPage[], expandedIds: string[]) =>
+  getRecursivePageList(
+    groupBy(pages, (page) => page.parentPageEntityId),
+    "null",
+    expandedIds,
+    0,
+  );
+
+export const getProjection = (
```

Comment:
Can you put a comment explaining what this does / is for? Probably a bit hard to follow for people not familiar with DND kit

Reply:
it's not so much about DND kit as it is about the Item Tree dragging solution. Added comments to the function in 2be59f165f2e986c4b13e26d41f7233f1d82a2e6